### PR TITLE
Launch basic Insiders page

### PR DESCRIPTION
**Pull request by ErikSchierboom**
- Build algorithm
- Add insiders_status field to users table

### CODE REVIEW

File: app/commands/insiders/activate.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class Insiders::Activate
+  include Mandate
+
+  initialize_with :user
+
+  def call
+    return unless user.insiders_status == :eligible
```
Comment (iHiD) [inline app/commands/insiders/activate.rb]: (This needs to pivot on the `lifetime_` stuff too)

File: app/commands/insiders/grant_access.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  include Mandate
+
+  initialize_with :user
+
+  def call
+    return if user.insiders_status == :lifetime_active
+    return if user.insiders_status == :active
+
+    user.update(insiders_status: :eligible)
+  end
```
Comment (iHiD) [inline app/commands/insiders/grant_access.rb]: Add a lock

File: app/commands/insiders/activate.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+    user.update(insiders_status: :active)
```
Comment (iHiD) [inline app/commands/insiders/activate.rb]: Add a lock to this and the lines above.

File: app/commands/user/add_roles.rb [hunk 4]
```diff
@@ -6,5 +6,6 @@ class User::AddRoles
   def call
     user.update(roles: user.roles + roles)
     User::SetDiscordRoles.(user)
+    User::InsidersStatus::TriggerUpdate.(user)
```
Comment (ErikSchierboom) [inline app/commands/user/add_roles.rb:9]: Now I need to remember to use the command when making someone a maintainer :)

File: app/commands/user/insiders_status/activate.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  def call
+    return unless %i[eligible eligible_lifetime].include?(user.insiders_status)
+
+    if user.insiders_status == :eligible
+      user.update(insiders_status: :active)
+    else
+      user.update(insiders_status: :lifetime_active)
+    end
+
+    User::Notification::Create.(user, :joined_insiders)
```
Comment (ErikSchierboom) [inline app/commands/user/insiders_status/activate.rb]: We probably want a different notification for lifetime insiders I guess?

File: app/commands/user/insiders_status/update.rb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  end
+
+  private
+  memoize
+  def eligibility_status = User::InsidersStatus::DetermineEligibilityStatus.(user)
+
+  def update_eligible_lifetime
+    return if user.insiders_status == :lifetime_active
+
+    user.update(insiders_status: :eligible_lifetime)
```
Comment (ErikSchierboom) [inline app/commands/user/insiders_status/update.rb]: I assume that if someone is currently active, we do want the status to become eligible lifetime? We do need to consider that to be "active" in the UI probably

File: app/commands/user/insiders_status/update.rb [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  def update_eligible
+    return if user.insiders_status == :lifetime_active
+    return if user.insiders_status == :active
+
+    user.update(insiders_status: :eligible)
+  end
+
+  def update_ineligible
+    return if user.insiders_status == :lifetime_active
+    return if user.insiders_status == :expired
```
Comment (ErikSchierboom) [inline app/commands/user/insiders_status/update.rb]: Expired should remain expired I think?

File: app/jobs/recheck_active_insiders_job.rb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+class RecheckActiveInsidersJob < ApplicationJob
+  queue_as :default
+
+  def perform
+    User.where(insiders_status: :active).find_each do |user|
```
Comment (ErikSchierboom) [inline app/jobs/recheck_active_insiders_job.rb]: Do we only want to check active insiders?

File: app/mailers/notifications_mailer.rb [hunk 9]
```diff
@@ -132,4 +132,13 @@ def automated_feedback_added
     subject = "There's new feedback on your solution to #{@track.title}/#{@exercise.title}"
     mail_to_user(@user, subject)
   end
+
+  def joined_insiders
+    notification = params[:notification]
+    @user = notification.user
+
+    @hide_header = true
+    subject = "Welcome to Exercism Insiders"
```
Comment (ErikSchierboom) [inline app/mailers/notifications_mailer.rb:141]: This is just a stub

File: db/migrate/20230412075814_add_insiders_status_to_users.rb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+class AddInsidersStatusToUsers < ActiveRecord::Migration[7.0]
+  def change
+    return if Rails.env.production?
+
+    add_column :users, :insiders_status, :tinyint, null: false, default: 0
```
Comment (ErikSchierboom) [inline db/migrate/20230412075814_add_insiders_status_to_users.rb:5]: Default is `unset` or `ineligible`?

File: app/models/user/notifications/joined_insiders_notification.rb [hunk 11]
```diff
@@ -0,0 +1,9 @@
+class User::Notifications::JoinedInsidersNotification < User::Notification
+  # TODO: determine what values to use
+  def url = Exercism::Routes.dashboard_url
+  def image_type; end
+  def image_url; end
+
+  # Users should only have this notification once
+  def guard_params = ""
```
Comment (ErikSchierboom) [inline app/models/user/notifications/joined_insiders_notification.rb:8]: Is this correct? As in: this prevents someone from re-receiving the notification?

File: app/commands/user/insiders_status/trigger_update.rb [hunk 12]
```diff
@@ -0,0 +1,10 @@
+class User::InsidersStatus::TriggerUpdate
+  include Mandate
+
+  initialize_with :user
+
+  def call
+    user.update(insiders_status: :unset)
```
Comment (ErikSchierboom) [inline app/commands/user/insiders_status/trigger_update.rb]: Don't unset active (or lifetime_active)

File: app/commands/user/insiders_status/update.rb (hunk 6 above)
Comment (iHiD) [reply app/commands/user/insiders_status/update.rb]: I've updated this. Check if you're happy.

File: app/commands/user/insiders_status/update.rb (hunk 7 above)
Comment (iHiD) [reply app/commands/user/insiders_status/update.rb]: I wonder if we actually need `expired`, or if we just need to change back to `illegible` and trigger an email when that happens.

File: app/jobs/recheck_active_insiders_job.rb (hunk 8 above)
Comment (iHiD) [reply app/jobs/recheck_active_insiders_job.rb]: Yes.

Comment (iHiD) [reply app/jobs/recheck_active_insiders_job.rb]: We could probably add a scope here for `User.insiders`

File: app/models/user/notifications/joined_insiders_notification.rb (hunk 11 above)
Comment (iHiD) [reply app/models/user/notifications/joined_insiders_notification.rb:8]: I think it should be `nil`. I also think we can move this into `app/models/concerns/is_paramaterised_sti.rb` under the `def i18n_params` line. If the tests still pass, I think we're fine to do this.

File: app/jobs/recheck_active_insiders_job.rb (hunk 8 above)
Comment (ErikSchierboom) [reply app/jobs/recheck_active_insiders_job.rb]: > User.insiders

That should include `lifetime_active` users too, and those we don't want to check here. Even `User.active_insiders` feels like it is not the right name.

File: app/commands/user/insiders_status/update.rb (hunk 7 above)
Comment (ErikSchierboom) [reply app/commands/user/insiders_status/update.rb]: I like it more when it is removed

File: app/commands/user/insiders_status/update.rb [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+  queue_as :default
+
+  initialize_with :user, :status_before_unset
+
+  def call
+    return unless user.insiders_status == :unset
+
+    user.with_lock do
+      case User::InsidersStatus::DetermineEligibilityStatus.(user)
```
Comment (iHiD) [inline app/commands/user/insiders_status/update.rb]: This need to be outside of the lock, as it's expensive.

File: app/commands/user/insiders_status/update.rb [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+      case eligibility_status
+      when :eligible_lifetime
+        update_eligible_lifetime
+      when :eligible
+        update_eligible
+      when :ineligible
+        update_ineligible
+      end
+    end
+  end
```
Comment (iHiD) [inline app/commands/user/insiders_status/update.rb:21]: This isn't essential now, but I'd quite to not include the notification sending bit within the lock.

Comment (ErikSchierboom) [reply app/commands/user/insiders_status/update.rb:21]: Follow-up PR to fix this: https://github.com/exercism/website/pull/3700

**Pull request merged**